src/canister/individual_user_template/src/api/cdao/airdrop.rs

```diff
@@ -0,0 +1,108 @@
+use candid::{Nat, Principal};
+use ic_base_types::PrincipalId;
+use ic_cdk::update;
+use ic_sns_root::pb::v1::{ListSnsCanistersRequest, ListSnsCanistersResponse};
+use icrc_ledger_types::icrc1::{account::Account, transfer::{Memo, TransferArg, TransferError}};
+use shared_utils::canister_specific::individual_user_template::types::{error::AirdropError, profile::UserProfileDetailsForFrontendV2};
+
+use crate::CANISTER_DATA;
+
+#[update]
+async fn request_airdrop(token_root: Principal, memo: Option<Memo>, amount: Nat, user_canister: Principal) -> Result<(), AirdropError> {
+    let current_caller = ic_cdk::caller();
+    let profile_info = get_profile_info(user_canister).await?;
+    
+    if profile_info.principal_id != current_caller {
+        return Err(AirdropError::CanisterPrincipalDoNotMatch);
+    }
+
+    if !is_airdrop_unclaimed(token_root, &current_caller)? {// assertion for token owner is checked here will return err if deployed sns cans not found
+        return Err(AirdropError::AlreadyClaimedAirdrop);
+    }
+
+    let amount = amount.min(1000u32.into());
+    if amount < 100u32 {
+        return Err(AirdropError::RequestedAmountTooLow)
+    }
+
+    set_airdrop_claiming(token_root, current_caller);
+
+    request_airdrop_internal(token_root, current_caller, memo, amount).await.inspect(|_|{
+        CANISTER_DATA.with_borrow_mut(|cans_data| {
```

Comment:
We are fetching token root canister twice by iterating over the array. Better to implement methods on DeployedCanisters

Reply:
Wouldn't it be better if i changed from` Vec<DeployedCdaoCanisters>` to `HashMap<TokenRootPrincipal, DeployedCdaoCanisters>` ? It would completely remove the iteration of the array and weird lifetime issues if i tried to get a mutable reference to the deployed_canister in a single iteration...

Reply:
aah maybe try to fetch it only once. It is not easy to change data structures in upgrade. You would require two upgrade if you decide to change the structure. One for moving the data from Vec to Hash and second to remove the upgrade bridge that you will write and it should run one after the other in order otherwise every upgrade would start failing. 

Better to keep it array and fetch it once.